Feature: hedge WebClient requests

Anusien opened the pull request:
Doing a preliminary pull request to get feedback before sending it to the Spring Commons folks.

Anusien committed:
Feature: hedge WebClient requests

Hedging is a mechanism for idempotent requests to automatically start a
new request after some time threshold has been reached. Unlike a simple
retry, hedging will use the first response returned.

Anusien commented:
TODO: Test the auto configuration

Anusien commented:
TODO: Make sure there are Javadocs with `@author` on all the files

Anusien commented:
TODO: Consider ripping the histogram stuff out of the hedging client to remove the direct dependency on micrometer

Anusien commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunction.java:
```diff
@@ -0,0 +1,113 @@
+ *
+ * {@see http://accelazh.github.io/storage/Tail-Latency-Study} for more background on how hedging works.
+ */
+public class HedgedRequestsExchangeFilterFunction implements ExchangeFilterFunction {
```
TODO: Make a generic version of this. I think it might be as easy as just taking in a `Mono` or a `Flux` (which for a filter could be `next.exchange(request)`.

Anusien replied:
It turns out this is more complicated because `HedgingClient#getDelayBeforeHedging` takes in the context of the `ClientRequest` in case people want to do something interesting there.

Anusien commented on spring-cloud-commons/src/test/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunctionTests.java:
```diff
@@ -0,0 +1,166 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
```
TODO: A test that verifies that the extra calls aren't made when one succeeds

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/Hedged.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.beans.factory.annotation.Qualifier;
+
+/**
+ * Annotation for a method invocation that is hedge-able. Hedging requests are a way to retry requests that have long
```
Nit: maybe "Hedging requests are a way" -> "Hedging requests is a way"

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactiveLoadBalancer.java:
```diff
@@ -46,7 +46,6 @@
 
 	@FunctionalInterface
 	interface Factory<T> {
-
```
Nit: they will probably tell you not to introduce whitespace changes

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/Hedged.java:
```diff
@@ -0,0 +1,62 @@
+@Documented
+@Inherited
+@Qualifier
+public @interface Hedged {
```
It would be nice to add documentation to https://github.com/spring-cloud/spring-cloud-commons/tree/master/docs/src/main/asciidoc about how to use this

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgingClient.java:
```diff
@@ -0,0 +1,36 @@
+
+	Duration getDelayBeforeHedging(ClientRequest request);
+
+	void record(ClientRequest request,
```
Do we need this here? It seems that `HedgingMetricsReporter` already has this

Anusien replied:
This is the feedback loop to get the latency of the request to update the 95th percentile.

When you originally implemented hedging, you did it via two separate `ExchangeFilterFunction` objects: one to record the latency, and one to calculate the 95th percentile and do the hedging. Putting this here is what enables this to become just a single function.

csabakos replied:
It does feel like it makes sense to separate getting the hedge parameters and recording. Some hedge policies might not even care about recording the latencies, maybe they just want to use fixed delay.

Anusien replied:
One of the problems with separating them was you have these two separate filters that have to be orchestrated in the right way. I think it's much cleaner to allow them to do their own tracking. If they don't want to, they just don't.

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/ReactiveHedgingAutoConfiguration.java:
```diff
@@ -0,0 +1,111 @@
+
+	@Nullable
+	public Object postProcessBeforeInitialization(@NonNull Object bean, String beanName) throws BeansException {
+		if (beanFactory == null || !(bean instanceof WebClient.Builder)) {
```
Nit: `beanFactory == null` is redundant

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/ReactiveHedgingAutoConfiguration.java:
```diff
@@ -0,0 +1,111 @@
+
+		HedgedRequestsExchangeFilterFunction hedgedFilterFunction = new HedgedRequestsExchangeFilterFunction(hedgingClient,
+				reporters);
+		builder.filter(hedgedFilterFunction);
```
Ordering is going to be a problem here. We need to figure out a way to control ordering in relation to other filter functions (and inject ourselves at the right place by default).

Anusien replied:
I think we just always want this at the end, right? So something like `@Order(Ordered.LOWEST_PRECEDENCE - 99)` (to put it right after circuit breaker and discovery?

Anusien replied:
Actually, it's not clear to me; should this be before or after load balancing?

csabakos replied:
I think it should be before load balancing because we want the hedged requests to go to a different service instance.

csabakos commented on spring-cloud-commons/src/test/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunctionTests.java:
```diff
@@ -0,0 +1,166 @@
+import org.springframework.web.reactive.function.client.ClientResponse;
+import org.springframework.web.reactive.function.client.ExchangeFunction;
+
+import static org.assertj.core.api.BDDAssertions.then;
```
Is this the right import?

Anusien replied:
This is the one they use all over the place. Their README makes me think this is important to them.

Anusien committed:
Ensure we don't create requests prematurely

We should not create another request until the first one has been
active for the appropriate amount of time. Unit tests showed the logic
to guarantee this was flawed. It's probably okay because we weren't
subscribing to those publishers, but still better to guarantee it this
way.

Anusien committed:
Verify that hedger autoconfiguration works

Implement a test to prove that the autoconfiguration works for the
hedger. Also add some minor cleanup and some javadocs.

Anusien committed:
Remove the @EnableHedger annotation

Reading the docs has made it clear to me this annotation is the old
approach. Instead we should just make it automatic based on the
presence of specific beans.

Anusien commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/Hedged.java:
```diff
@@ -0,0 +1,62 @@
+	 * {@link HedgingMetricsReporter} and they will get called on each request.
+	 * @return the hedged reporter bean names
+	 */
+	String[] metricReporters() default {};
```
Should we move the metrics reporters to the `HedgingClient`? I think not (because it makes them way more complicated), but I want to ask.

csabakos replied:
Agree that it's probably better to keep them separate

Anusien committed:
Document request hedging in the asciidocs

Attempting to document request hedging showed how cumbersome the name
and constructor was for the "default" HedgingClient. I renamed it
and dramatically simplified the constructor.

Anusien committed:
Fixes some minor spacing on the asciidocs

Anusien committed:
Fixes accidental changes

Doing an initial review of my changes, I discovered a few places I
accidentally introduced imports or whitespace changes. This reverses
those.

Anusien committed:
Fixes unit tests for LatencyPercentileHedgingClient

A previous commit reordered the constructor. This fixes the methods
to call it properly.

csabakos commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -1059,3 +1059,99 @@ myrandom=${cachedrandom.appname.value}
+    @Bean
+    HedgingClient hedgingClient() {
+    	return new LatencyPercentileHedgingClient(
+            3, // Maximum extra requests to execute
```
Suggestion: let's create a builder for `LatencyPercentileHedgingClient` so that these comments can turn into self-documenting setter methods.

Anusien replied:
That feels like too much complication for this thing. I'm not sure this client is the right way to do it anyway.

csabakos commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -1059,3 +1059,99 @@ myrandom=${cachedrandom.appname.value}
+
+    @Bean
+    HedgingMetricsReporter hedgingReporterA() {
+		new HedgingMetricsReporter() {
```
I think a `return` is missing but maybe it would be even better to use a lambda

Anusien replied:
I saw others in the file that do it this way.

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgingClient.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @author Kevin Binswanger
+ */
+public interface HedgingClient {
```
I don't think `HedgingClient` is a great name because this is not really a client of anything... maybe `HedgePolicy`? I'll let the Spring folks suggest a better name :)

Anusien replied:
I think `HedgingPolicy` is a much better name.

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgingClient.java:
```diff
@@ -0,0 +1,39 @@
+ * @author Kevin Binswanger
+ */
+public interface HedgingClient {
+	boolean shouldHedge(ClientRequest request);
```
It might make sense to combine these 3 methods into something like:
`ClientRequestHedgeConfiguration getHedgeConfiguration(ClientRequest request)`

Anusien replied:
That requires making this extra object, the configuration. I'll think about it because it makes the ExchangeFilterFunction a slight bit simpler.

Anusien replied:
I do think we could probably get rid of `shouldHedge` and replace it with `getNumberOfHedgedRequests` returning 0.

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgingMetricsReporter.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.web.reactive.function.client.ClientResponse;
+
+/**
+ * Hedging could create multiple requests to a given service. Any monitoring tools put around the request will therefore
```
`Any monitoring tools put around the request will therefore see a bunch of requests`

I think this depends on where you put the hedging exchange filter in the chain: outside or inside of the metrics reporting filter

Anusien committed:
Fixes the syntax of the example hedging Metrics Reporter beans

Anusien commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunction.java:
```diff
@@ -0,0 +1,114 @@
+	private Mono<ClientResponse> withSingleMetricsReporting(
+			ClientRequest request,
+			Mono<ClientResponse> response,
+			@Nullable Integer hedgeNumber
```
Right now `hedgeNumber` is `null` for not hedged, and `1` for the first hedge. Should this just be `0` for the first request instead?

onobc replied:
I like nonnull and 0

Anusien commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/LatencyPercentileHedgingClient.java:
```diff
@@ -0,0 +1,119 @@
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Predicate;
+
+import io.micrometer.core.instrument.Clock;
```
I just double checked. `io.micrometer.core` is available transitively through an optional dependency (`org.springframework.boot:spring-boot-starter-actuator`). I kind of want to split this thing out, but I'm not sure the best way to do it. Guava and Apache both have some methods to do quantiles, but they're not optimized for a moving dataset like micrometer. I see a couple of options:

1.  Keep it mostly as-is. Add micrometer as an optional dependency (to be explicit) and indicate that you need to provide micrometer if you use the LatencyPercentileHedgingClient.
2.  Pull micrometer in directly, leave everything else as-is.
3.  Adopt some other histogram mechanism.
4.  Move the LatencyPercentileHedgingClient to a separate package, the way the load-balancer does.

I think 4 is probably the best option?

csabakos replied:
4 sounds good but it might be best to ask the Spring Cloud maintainers

Anusien committed:
Rename HedgingClient to HedgingPolicy

Reviewers pointed out that this isn't actually a client of anything.

Anusien committed:
Completes renaming HedgingClient to HedgingPolicy

Anusien committed:
Complete the hedgingPolicy rename

Some things somehow got named to heckingPolicy instead. This fixes that.

onobc commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -1059,3 +1059,99 @@ myrandom=${cachedrandom.appname.value}
+    @Autowired
+    private WebClient.Builder webClientBuilder;
+
+    public String doOtherStuff() {
```
`public Mono<String> doOtherStuff() {`

onobc commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -1059,3 +1059,99 @@ myrandom=${cachedrandom.appname.value}
+    @Hedged(hedgingPolicy = "hedgingPolicy")
+    @Bean
+    WebClient.Builder webClientBuilder() {
+        return WebClient.builder();
```
So the `org.springframework.boot.autoconfigure.web.reactive.function.client.WebClientAutoConfiguration` creates a builder for us that respects the available WebClientCustomizer. The expected use is to use the provided builder but create customizers to extend/augment the provided builder. 
```java
    @Bean
    @Scope("prototype")
    @ConditionalOnMissingBean
    public Builder webClientBuilder(ObjectProvider<WebClientCustomizer> customizerProvider) {
        Builder builder = WebClient.builder();
        customizerProvider.orderedStream().forEach((customizer) -> {
            customizer.customize(builder);
        });
        return builder;
    }
```

So we would either need to at least do the same in this example (although that will steer away from the provided builder) or move this mechanism to a WebClientCustomizer.

csabakos replied:
Hmm maybe it's easier/better to get rid of the `@Hedged` annotation and do something like this instead:
```
@Bean
public WebClientCustomizer hedgingWebClientCustomizer(final HedgingPolicy hedgingPolicy) {
    return HedgingWebClientCustomizer.of(hedgingPolicy);
}
```

onobc commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -1059,3 +1059,99 @@ myrandom=${cachedrandom.appname.value}
+====
+[source,java,indent=0]
+----
+import org.springframework.cloud.client.hedger.HedgingMetricsReporter;@Configuration
```
Nit: same line formattting?

Anusien replied:
The stupid editor keeps trying to fix imports in this code and it doesn't even put the line break there. :frowning_face:

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/Hedged.java:
```diff
@@ -0,0 +1,62 @@
+ * 95th percentile latency in exchange for 2-5% traffic increase. (The actual numbers depend on the distribution of
+ * latencies.)
+ *
+ * {@see http://accelazh.github.io/storage/Tail-Latency-Study} for more background on how hedging works.
```
Possibly better suited in the doc. Also is it ok to cite this study? Probably fine, but should check any restrictions on it.

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunction.java:
```diff
@@ -0,0 +1,114 @@
+	private List<HedgingMetricsReporter> hedgingMetricsReporterList;
+
+	public HedgedRequestsExchangeFilterFunction(
+			HedgingPolicy hedgingPolicy, List<HedgingMetricsReporter> hedgingMetricsReportersList) {
```
If `hedgingMetricsReportersList` can be null, may want to add `@org.springframework.lang.Nullable`

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunction.java:
```diff
@@ -0,0 +1,114 @@
+				.elapsed()
+				.doOnSuccess(tuple -> {
+					hedgingPolicy.record(request, tuple.getT2(), tuple.getT1(), hedgeNumber);
+					hedgingMetricsReporterList.forEach(reporter -> reporter.record(request, tuple.getT2(), tuple.getT1(), hedgeNumber));
```
What happens when `reporter.record` throws runtime exception?

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgingMetricsReporter.java:
```diff
@@ -0,0 +1,44 @@
+	 * @param response The eventual HTTP response.
+	 * @param elapsedMillis The number of milliseconds elapsed just for this attempt (ignores time spent waiting on
+	 *                      previous attempts).
+	 * @param hedgeNumber {null} for the original request, or which hedge attempt this is (starting at 1).
```
After reading this, I really think that 0 should be the 1st one.

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunction.java:
```diff
@@ -0,0 +1,114 @@
+				.mergeWith(
+						Flux.range(1, numHedges)
+								.delayElements(delay)
+								.flatMap(hedgeNumber -> withSingleMetricsReporting(request, next.exchange(request), hedgeNumber)
```
So in multiple places we say that we only call report for the actual request that we use (1st to succeed). Refresh my memory how this code does not possibly call if multiple of the hedges succeed?

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgingPolicy.java:
```diff
@@ -0,0 +1,51 @@
+public interface HedgingPolicy {
+	boolean shouldHedge(ClientRequest request);
+
+	int getNumberOfHedgedRequests(ClientRequest request);
```
Maybe rename this a bit as it sounds like "the number of requests that were hedged". Maybe `getMax..` or `getTotal..`

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/LatencyPercentileHedgingPolicy.java:
```diff
@@ -0,0 +1,119 @@
+ * @author Csaba Kos
+ * @author Kevin Binswanger
+ */
+public class LatencyPercentileHedgingPolicy implements HedgingPolicy {
```
Some javadoc would be helpful here too.

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/LatencyPercentileHedgingPolicy.java:
```diff
@@ -0,0 +1,119 @@
+ */
+public class LatencyPercentileHedgingPolicy implements HedgingPolicy {
+	private final Predicate<ClientRequest> shouldTrackRequestPredicate;
+	private final int numHedgedRequests;
```
Should we provide an ability to report on, or answer what the current histogram is looking like at any point in time?

onobc commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/ReactiveHedgingAutoConfiguration.java:
```diff
@@ -0,0 +1,87 @@
+		}
+
+		@Nullable
+		public Object postProcessBeforeInitialization(@NonNull Object bean, String beanName) throws BeansException {
```
I think we can do this via WebClientCustomizer. It could be an auto-configured bean that you just wire in the HedgingPolicy and the reporters on w/o the need to do the `@Hedged` annotation.

Anusien replied:
Interestingly, all the other things in Spring Cloud Commons use `BeanPostProcessor`. I'm guessing this is because `WebClientCustomizer` is the newer way to do it, and a lot of this stuff is old.

I think I actually want the `@Hedged` annotation though. I was looking at the different way `@LoadBalanced` is configured, and how more flexible and customizable their auto configuration is. The difference is that I think you want the ability to set a different `HedgedPolicy` on each `WebClient.Builder`.

Well that's what I thought, and then I saw that `LoadBalanced` actually allows this too. I'm gonna make some changes.

csabakos replied:
I'm guessing that they use `BeanPostProcessor` because nobody is forcing the user to use the default `WebClient.Builder` (the one that applies the customizers)... but not sure

onobc commented on spring-cloud-commons/src/test/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunctionTests.java:
```diff
@@ -0,0 +1,215 @@
+		verify(metricsReporter).record(clientRequest, response, 50, null);
+		verify(exchangeFunction, times(1)).exchange(clientRequest);
+	}
+
```
What about the case where multiple (or all succeed)?

Anusien replied:
What would we asset in that case?

onobc commented on spring-cloud-commons/src/test/java/org/springframework/cloud/client/hedger/ReactiveHedgingAutoConfigurationTests.java:
```diff
@@ -0,0 +1,106 @@
+public class ReactiveHedgingAutoConfigurationTests {
+
+	@Autowired
+	private WebClient.Builder webClientBuilder;
```
I would add 2 negative cases, where WebClient.class not available and where HedgingPolicy bean not defined and verify the auto-config does not kick in. I would use `ApplicationContextRunner` for this. See org.springframework.cloud.client.ReactiveCommonsClientAutoConfigurationTests#worksWithoutActuator for example.

Anusien committed:
Specifies HedgerPolicy via factory

The mechanism of specifying a HedgerPolicy via a parameter of the
Hedger annotation was clumsy and error-prone. Looking at the
LoadBalancer code suggested another way: accepting a factory that could
be run for each ClientRequest.

Also renamed Hedging to Hedger to match other packages here.

Anusien committed:
Clarify that HedgerListeners run on each request

I previously wrote (incorrectly) that the point of a HedgerListener was
to get called just on the final request. This is incorrect. It gets
called on every request. Clarified this in the documentation and
outlined the use case.

Anusien committed:
Fixes some hedging syntax errors in the documentation

csabakos commented on spring-cloud-commons/src/main/java/org/springframework/cloud/client/hedger/HedgerListener.java:
```diff
@@ -0,0 +1,47 @@
+	 * @param response The eventual HTTP response.
+	 * @param elapsedMillis The number of milliseconds elapsed just for this attempt (ignores time spent waiting on
+	 *                      previous attempts).
+	 * @param requestNumber Which attempt this is. The original will be 0, and the first hedged request will be 1.
```
I believe originally this was `@Nullable` so that the method can tell if hedging is enabled for the request or not.

Currently it cannot tell if hedging is disabled, or hedging is enabled but it's the original request.

Anusien replied:
oh, hrm.
